virttest/libvirt_vm.py

```diff
@@ -281,6 +281,97 @@ def is_autostart(self):
+                virsh.suspend(self.name, uri=self.connect_uri, debug=debug)
+            elif state == "shutoff":
+                if shutoff_method == "shutdown":
+                    virsh.shutdown(self.name, uri=self.connect_uri, debug=debug)
+                elif shutoff_method == "destroy":
+                    virsh.destroy(self.name, uri=self.connect_uri, debug=debug)
+                else:
+                    logging.error("Invalid shutoff_method %s", shutoff_method)
+            elif state == "pmsuspended":
+                virsh.dompmsuspend(self.name, "mem", uri=self.connect_uri, debug=debug)
+        elif cur_state == "paused":
+            if state == "running":
+                virsh.resume(self.name, uri=self.connect_uri, debug=debug)
+            elif state == "os-booted":
+                virsh.resume(self.name, uri=self.connect_uri, debug=debug)
+                if not self.login():
+                    logging.error("Fail to loggin %s when set VM state from %s to %s",
+                                  self.name, cur_state, state)
+            elif state == "shutoff":
+                virsh.destroy(self.name, uri=self.connect_uri, debug=debug)
+            elif state == "pmsuspended":
+                virsh.resume(self.name, uri=self.connect_uri, debug=debug)
+                virsh.dompmsuspend(self.name, "mem", uri=self.connect_uri, debug=debug)
+        elif cur_state == "shutoff":
+            if state == "running":
+                virsh.start(self.name, uri=self.connect_uri, debug=debug)
+            elif state == "os-booted":
+                virsh.start(self.name, uri=self.connect_uri, debug=debug)
+                if not self.wait_for_login():
+                    logging.error("Fail to loggin %s when set VM state from %s to %s",
+                                  self.name, cur_state, state)
+            elif state == "paused":
+                virsh.start(self.name, '--paused', uri=self.connect_uri, debug=debug)
+            elif state == "pmsuspended":
+                virsh.start(self.name, uri=self.connect_uri, debug=debug)
+                virsh.dompmsuspend(self.name, "mem", uri=self.connect_uri, debug=debug)
+        elif cur_state == "pmsuspended":
+            if state == "running":
+                virsh.dompmwakeup(self.name, uri=self.connect_uri, debug=debug)
+            elif state == "os-booted":
+                virsh.dompmwakeup(self.name, uri=self.connect_uri, debug=debug)
+                if not self.wait_for_login():
+                    logging.error("Fail to loggin %s when set VM state from %s to %s",
+                                  self.name, cur_state, state)
+            elif state == "paused":
+                virsh.dompmwakeup(self.name, uri=self.connect_uri, debug=debug)
+                virsh.suspend(self.name, uri=self.connect_uri, debug=debug)
+            elif state == "shutoff":
+                virsh.destroy(self.name, uri=self.connect_uri, debug=debug)
+
```

Comment:
How to make sure that the virsh commands don't fail when executing them? How about we set `ignore_status=False`?
Or we could add state checking after setting part is done, maybe a `wait_for(vm state is xxx)` ?

Reply:
If we set `ignore_status=False`, the test code would have to test if the target state would have to test if the expected target state was actually established, shouldn't that be the responsibility of this very method `set_state`?

IMO, it makes sense to use `wait_for`, e.g. the pattern
```
wait_for(vm state is xxx, ignore_error=True, ...)
if vm.state() != expected_state:
 raise AssertionError("Couldn't establish state %s....)
```